buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt

```diff
@@ -200,12 +201,19 @@ class UnitTestAndCompilePlugin : Plugin<Project> {
         inputs.property("javaInstallation", Callable { jvmForTest.vendorAndMajorVersion })
     }
 
+    private
+    fun Test.addOsAsInputs() {
+        // https://github.com/gradle/gradle-private/issues/2831
+        inputs.property("operatingSystem", "${OperatingSystem.current().name} ${System.getProperty("os.arch")}")
+    }
+
     private
     fun Project.configureTests() {
         tasks.withType<Test>().configureEach {
             maxParallelForks = project.maxParallelForks
 
             configureJvmForTest()
+            addOsAsInputs()
```

Comment:
That configures the tests and the distribution tests. Should we maybe only add the property to unit tests since the `DistributionTest`s declare the OS already?

Using the runtime API makes the property name clash with the one declared in `DistributionTest`, I think we then track the one added via the runtime API.

Reply:
How about remove the declaration from `DistributionTest` and use runtime API here for both unit tests and integration tests? This can make the configuration in one place, instead of two places.

Reply:
Sounds good to me. Please add a comment why we need the OS as an input for unit tests as well.